### ISSUE TITLE
Invalid JSON output in compiler artifacts with multiple contracts with the same name

### ISSUE THREAD

**Issue by spalladino** [OpenZeppelin/openzeppelin-sdk#1203]
Given a project with more than one contract with the same name (see [here](https://github.com/crazyrabbitLTC/ETHBERLIN/tree/93592ea49f4f908ac64eda9a49e6095a3b31d064) for an example), the compiler generates invalid JSON files for them.

```
  "compiler": {
    "name": "solc",
    "version": "0.5.3+commit.10d17f24.Emscripten.clang",
    "optimizer": {},
    "evmVersion": "constantinople"
  }
}
nople"
  }
}
```

This is probably because we have more than one stream writing to the same file. To fix this, we should check the list of compilerOutput and remove any duplicate contract names before writing.

https://github.com/OpenZeppelin/openzeppelin-sdk/blob/45778202249e725a368fdb6a6600ccd49e1c66bb/packages/cli/src/models/compiler/solidity/SolidityProjectCompiler.ts#L189-L194

However, having duplicated contract names can yield to other errors, such as invalid storage layout calculations (since we rely on the AST of the build artifacts for that) or even unexpected contracts getting created (since the user may expect one of the bytecodes to be deployed, but we actually pick the other). We should also rise a big warning if there are indeed duplicated contracts.

**Comment by melnikaite** [comment]
Could you please update link with example?

**Comment by melnikaite** [comment]
I found 2 serious issues

First of all flat list of artefacts, in case we have Counter.sol and subdir/Counter.sol in build/contracts we will have artefact only for subdir/Counter.sol, so we are missing artefact for Counter.sol

And second issue you actually faced with is having multiple contracts in the same .sol file, e.g.
```
pragma solidity 0.5.4;

contract Counter {
    uint256 public value;

    function increase() public {
        value++;
    }

    function decrease() public {
        value--;
    }
}

contract X {}
```

For now the only workaround I see is keeping each contract in separate file and use unique contract names. However you still can use subfolders for imported contracts.

**Comment by nventuro** [comment]
After encountering this myself, I'd say that erroring out early is the way to go: many parts of the SDK assume contract names are unique (e.g. when deploying via `create`), so we should notify the user as soon as we detect this and abort.

**Comment by abcoathup** [comment]
Also reported in the community forum:
https://forum.openzeppelin.com/t/error-unexpected-string-in-json-at-position-23357/1614

I reproduced (and went to create an issue, and then recalled having seen this issue, oops):

Create an OpenZeppelin SDK project
```
mkdir test
cd test
npm init -y
oz init
```

Create two contract files `A.sol` and `B.sol` with one contract larger than the other.  

Renaming the contracts so they have the same name e.g. `contract A` and running `oz compile` results in an `A.json` which appears to have the long contract and short contract in the same json.

Running `oz create` then gives the error:
```
$ oz create
test/build/contracts/A.json: Unexpected token r in JSON at position 1620
```
I haven't tracked down the exact order of editing and saving to reproduce every time.

## A.sol
```
pragma solidity ^0.5.0;

contract A {
  uint256 public value;
}
```

## B.sol
```
pragma solidity ^0.5.0;

contract A {
}
```
## A.json
Corruption
```
{
  "fileName": "B.sol",
  "contractName": "A",
  "source": "pragma solidity ^0.5.0;\n\ncontract A {\n}",
  "sourcePath": "contracts/B.sol",
  "sourceMap": "25:14:1:-;;;;8:9:-1;5:2;;;30:1;27;20:12;5:2;25:14:1;;;;;;;;;;",
  "deployedSourceMap": "25:14:1:-;;;;;;;;;;;",
  "abi": [],
  "ast": {
    "absolutePath": "contracts/B.sol",
    "exportedSymbols": {
      "A": [
        7
      ]
    },
    "id": 8,
    "nodeType": "SourceUnit",
    "nodes": [
      {
        "id": 6,
        "literals": [
          "solidity",
          "^",
          "0.5",
          ".0"
        ],
        "nodeType": "PragmaDirective",
        "src": "0:23:1"
      },
      {
        "baseContracts": [],
        "contractDependencies": [],
        "contractKind": "contract",
        "documentation": null,
        "fullyImplemented": true,
        "id": 7,
        "linearizedBaseContracts": [
          7
        ],
        "name": "A",
        "nodeType": "ContractDefinition",
        "nodes": [],
        "scope": 8,
        "src": "25:14:1"
      }
    ],
    "src": "0:39:1"
  },
  "bytecode": "0x608060405234801560105760006000fd5b506015565b60478060226000396000f3fe608060405260043610600c575b60006000fdfea265627a7a72315820f6c84d2e97e03a7a177b2285452560c099a8ff0a7521565d01cab2741bf4702e64736f6c634300050c0032",
  "deployedBytecode": "0x608060405260043610600c575b60006000fdfea265627a7a72315820f6c84d2e97e03a7a177b2285452560c099a8ff0a7521565d01cab2741bf4702e64736f6c634300050c0032",
  "compiler": {
    "name": "solc",
    "version": "0.5.12+commit.7709ece9.Linux.g++",
    "optimizer": {},
    "evmVersion": "constantinople"
  }
}
riable": true,
            "storageLocation": "default",
            "typeDescriptions": {
              "typeIdentifier": "t_uint256",
              "typeString": "uint256"
            },
            "typeName": {
              "id": 2,
              "name": "uint256",
              "nodeType": "ElementaryTypeName",
              "src": "40:7:0",
              "typeDescriptions": {
                "typeIdentifier": "t_uint256",
                "typeString": "uint256"
              }
            },
            "value": null,
            "visibility": "public"
          }
        ],
        "scope": 5,
        "src": "25:38:0"
      }
    ],
    "src": "0:63:0"
  },
  "bytecode": "0x60806040523480156100115760006000fd5b50610017565b608c806100256000396000f3fe608060405234801560105760006000fd5b5060043610602c5760003560e01c80633fa4f24514603257602c565b60006000fd5b6038604e565b6040518082815260200191505060405180910390f35b6000600050548156fea265627a7a723158206e8656af342f560118938ca66f3e1d7ae5fb7f1019e2f68b3947985765c0563c64736f6c634300050c0032",
  "deployedBytecode": "0x608060405234801560105760006000fd5b5060043610602c5760003560e01c80633fa4f24514603257602c565b60006000fd5b6038604e565b6040518082815260200191505060405180910390f35b6000600050548156fea265627a7a723158206e8656af342f560118938ca66f3e1d7ae5fb7f1019e2f68b3947985765c0563c64736f6c634300050c0032",
  "compiler": {
    "name": "solc",
    "version": "0.5.12+commit.7709ece9.Linux.g++",
    "optimizer": {},
    "evmVersion": "constantinople"
  }
}

```

**Comment by frangio** [comment]
>I'm facing this issue (invalid artifacts) in a setup where I have no duplicated contracts.

The duplicate was probably behind an `import`.

**Comment by frangio** [comment]
Fixed in #1411. Will be in master after #1438 is merged.
